autopilot/contractor/contractor.go

```diff
@@ -1233,56 +1049,57 @@ func performContractMaintenance(ctx *mCtx, alerter alerts.Alerter, bus Bus, chur
 		return false, err
 	}
 
-	// fetch old set
-	oldSet, err := bus.Contracts(ctx, api.ContractsOpts{ContractSet: ctx.ContractSet()})
-	if err != nil && !utils.IsErr(err, api.ErrContractSetNotFound) {
-		return false, fmt.Errorf("failed to fetch old contract set: %w", err)
+	// STEP 4: update contract usability
+	oldGC, err := bus.Contracts(ctx, api.ContractsOpts{FilterMode: api.ContractFilterModeGood})
+	if err != nil {
+		return false, fmt.Errorf("failed to fetch good contracts: %w", err)
 	}
-
-	// merge kept and formed contracts into new set
-	newSet := make([]api.ContractMetadata, 0, len(keptContracts)+len(formedContracts))
-	newSet = append(newSet, keptContracts...)
-	newSet = append(newSet, formedContracts...)
-
-	// STEP 4: update contract set
-	if err := updateContractSet(ctx, bus, oldSet, newSet); err != nil {
-		return false, err
+	newGC := make([]api.ContractMetadata, 0, len(keptContracts)+len(formedContracts))
+	newGC = append(newGC, keptContracts...)
+	newGC = append(newGC, formedContracts...)
+	updated, err := updateContractUsability(ctx, bus, alerter, oldGC, newGC, reasons, logger)
+	if err != nil {
+		return false, fmt.Errorf("failed to update contracts: %w", err)
 	}
 
-	// STEP 5: perform minor maintenance such as cleanups and broadcasting
-	// revisions
+	// STEP 5: perform post maintenance tasks
 	if err := performPostMaintenanceTasks(ctx, bus, alerter, cc, rb, logger); err != nil {
-		return false, err
+		return updated, err
 	}
-
-	// STEP 6: log changes and register alerts
-	return computeContractSetChanged(ctx, alerter, bus, churn, logger, oldSet, newSet, churnReasons)
+	return updated, nil
 }
 
-func updateContractSet(ctx *mCtx, bus Bus, oldSet, newSet []api.ContractMetadata) error {
-	var newSetIDs []types.FileContractID
-	for _, contract := range newSet {
-		newSetIDs = append(newSetIDs, contract.ID)
-	}
-	inOldSet := make(map[types.FileContractID]struct{})
-	for _, c := range oldSet {
-		inOldSet[c.ID] = struct{}{}
+func updateContractUsability(ctx *mCtx, bus Bus, alerter alerts.Alerter, oldGC, newGC []api.ContractMetadata, reasons map[types.FileContractID]string, logger *zap.SugaredLogger) (bool, error) {
+	wasGC := make(map[types.FileContractID]struct{})
+	for _, c := range oldGC {
+		wasGC[c.ID] = struct{}{}
 	}
-	var toAdd []types.FileContractID
-	for _, c := range newSet {
-		if _, ok := inOldSet[c.ID]; !ok {
-			toAdd = append(toAdd, c.ID)
+
+	var cug, cub []types.FileContractID
```

Comment:
Please use slightly more descriptive variable names. I'm not sure I'll ever figure out what these mean^^
e.g. `badToGood` and `goodToBad` and for `oldGC`, `newGC` and `wasGC` just `oldGood`, `newGood` and `wasGood`.